packages/react-jss/src/index.d.ts

```diff
@@ -72,26 +76,25 @@ interface CreateUseStylesOptions<Theme = DefaultTheme> extends BaseOptions<Theme
   name?: string
 }
 
-declare function createUseStyles<Theme = DefaultTheme, C extends string = string>(
-  styles: Styles<C> | ((theme: Theme) => Styles<C>),
+declare function createUseStyles<Props = unknown, Theme = DefaultTheme, C extends string = string>(
```

Comment:
Yeah, this one is a bit more difficult and is why I asked the question about ordering the generic type parameters.

Having `props` as the third argument requires people to "reach over" `Theme` and explicitly define `ClassNames` in order to provide any kind of prop info. I'm not sure how intuitive it is.

If people use theme and class names significantly more than props, then putting props last would make more sense. (This is my first time really getting into the TS side of React JSS, so I don't know what's commonly used and what isn't.) I'm trying to think of how intuitive things will be moving forward too. If it's something unintuitive, I imagine it'll be stuck there for quite some time.

Reply:
Agreed, this is a tricky situation until microsoft/TypeScript#26349 is merged. There are different approaches to using JSS - we use theme very often but rarely use props so I think my opinion here is biased. 

My suggestion would be to make `props` the third argument so that this doesn't break anything, but I think that's for @kof to decide.

Reply:
Just so it's visible when @kof responds:

Looking to the future, I think I'd prefer `<Name, Props, Theme>` (or `<Name, Theme, Props>`, depending on which one users employ more frequently). For users, I think that would be significantly more intuitive. Note that if possible, I'd prefer for `Styles`, `withStyles`, and `createUseStyles` to have the same type parameter order. This would improve consistency and minimize confusion.

The overarching question(s) is/are: What is the most intuitive ordering of the generic parameters that minimizes friction? Will many complain about an unintuitive ordering of parameters in the future, especially after `Props` finally gets type support? How difficult will introducing something more intuitive get in the future? Is it worth a breaking change?

In my (limited) opinion, I feel that long-term, something more intuitive is worth the change. This is riding on the assumption that the breaking change will not be too difficult for users to fix, and that it will be very difficult (if not impossible) to introduce a more intuitive structure in the future if we neglect it now. This also rides on the reality that after microsoft/TypeScript#26349 is merged, we'll be virtually unable to take practical advantage of it if `Names` is not the first type parameter. (Reaching over everything to put a `_` on the last parameter will make things ugly for the user.)

In the mean time, for hooks users who _only_ use theme, people can get `Theme` typing implicitly by using the function version of `createUseStyles`, especially with the help of #1453. So they wouldn't have to "reach over" `Name` or `Props`. (Note also that `Props` lacks a "non-hackish" equivalent if it is placed last.) Honestly, since for current users `Name` loses implicit typing when `Theme` is supplied, I'd wager that most people would prefer to move to the implicit types anyway. (That's an _assumption_ though. And I don't know how big migrating pains would be.)

Migration docs can always be added as well. Whatever the decision, I'm happy to change the ordering of the generic type params again once a conclusion is reached.

Reply:
So the question is if people are gonna have styles with props only vs props and theme vs. theme only more often? I think I would put theme last, because conditional styles is probably a more popular case than theme based once.

I think people should really put theming into the ui library components which would be reused everywhere, but application level styles should rarely need to access theme.

Reply:
@ITenthusiasm - LGTM, provided @kof is happy with this being a breaking change. 

I think we'll probably need some sort of upgrade guide for typescript users to explain what all the new errors mean (due to the change in argument order)

Reply:
Agreed! I was thinking of creating some kind of RFC issue related to the docs for that after this got merged.